### PR TITLE
wolfSSHd Wildcard Config Includes

### PR DESCRIPTION
**Pull request by ejohnstown**
1. Fix the wildcard config file include.
2. Update the guard flags so macOS can also use wildcards.
3. Change the user priviledge separating setting to a bitfield.
4. Add test_configuration test to gitignore.
5. Process the included config files in strcmp order.

### CODE REVIEW

**Review by haydenroche5 — CHANGES_REQUESTED** [review]
Is it feasible to add tests to test_configuration.c for this?

**Comment by haydenroche5** [comment]
Can we use the X/W functions (e.g. WREADDIR as opposed to readdir) throughout the Include stuff?

**Review by JacobBarthelmeh — APPROVED** [review]
The added test requires --enable-sftp on (./configure --enable-sshd without sftp gets compile errors). Otherwise looks great.

File: apps/wolfsshd/configuration.c [hunk 1]
```diff
@@ -503,49 +505,80 @@ static int HandleInclude(WOLFSSHD_CONFIG *conf, const char *value)
             prefixLen = (int)(ptr - value);
         }
 
-        d = opendir(path);
-        if (d) {
-            while ((dir = readdir(d)) != NULL) {
-                if (dir->d_type == DT_DIR) {
-                    /* Skip sub-directories */
-                    continue;
+        if (!WOPENDIR(NULL, conf->heap, &d, path)) {
```
Comment (haydenroche5) [inline apps/wolfsshd/configuration.c:508]: Let's change the `memcpy`s above to `WMEMCPY`s. Also move the `} else {` onto a new line.

File: apps/wolfsshd/configuration.c [hunk 2]
```diff
@@ -503,49 +505,80 @@ static int HandleInclude(WOLFSSHD_CONFIG *conf, const char *value)
...
-                        if ((int)WSTRLEN(dir->d_name) <= prefixLen) {
-                            continue;
-                        }
-                        if (WSTRNCMP(dir->d_name, prefix, prefixLen) != 0) {
-                            continue;
+            }
+            WREWINDDIR(&d);
+
+            if (fileCount > 0) {
+                fileNames = (char**)WMALLOC(fileCount * sizeof(char*), NULL, 0);
```
Comment (haydenroche5) [inline apps/wolfsshd/configuration.c:522]: Can use `conf->heap` for the heap argument to `WMALLOC` here and elsewhere.

File: apps/wolfsshd/configuration.c [hunk 3]
```diff
@@ -474,10 +475,11 @@ static int HandleInclude(WOLFSSHD_CONFIG *conf, const char *value)
...
     if (found) {
-#ifdef __unix__
+#if defined(__unix__) || defined(__unix) || \
+    (defined(__APPLE__) && defined(__MACH__))
         int ret;
         struct dirent *dir;
-        DIR *d;
+        WDIR d;
         char *path;
         char *filepath = (char*)WMALLOC(PATH_MAX, NULL, 0);
```
Comment (haydenroche5) [inline apps/wolfsshd/configuration.c:484]: We need to check the return of this `WMALLOC`.

**Review by haydenroche5 — APPROVED** [review]

File: apps/wolfsshd/configuration.c (hunk 3 above)
Comment (ejohnstown) [reply apps/wolfsshd/configuration.c:484]: A little later, the path variable is also malloced without a check, and isn't freed. Fixing.

File: apps/wolfsshd/configuration.c (hunk 2 above)
Comment (ejohnstown) [reply apps/wolfsshd/configuration.c:522]: And added dynamic types.

File: apps/wolfsshd/configuration.c (hunk 1 above)
Comment (ejohnstown) [reply apps/wolfsshd/configuration.c:508]: Check

File: apps/wolfsshd/configuration.c [hunk 4]
```diff
@@ -490,69 +496,114 @@ static int HandleInclude(WOLFSSHD_CONFIG *conf, const char *value)
...
+                    WSNPRINTF(filepath, PATH_MAX, "%s/%s", path,
+                             fileNames[i]);
+                }
+                ret = wolfSSHD_ConfigLoad(conf, filepath);
+                if (ret != WS_SUCCESS) {
+                    WCLOSEDIR(&d);
+                    WFREE(fileNames, conf->heap, DYNTYPE_PATH);
+                    WFREE(filepath, conf->heap, DYNTYPE_PATH);
+                    WFREE(path, conf->heap, DYNTYPE_PATH);
+                    return ret;
```
Comment (haydenroche5) [inline apps/wolfsshd/configuration.c:592]: Would prefer this function have a single return at the end, which'll save us having to do all the freeing at multiple different return sites, but not critical.

File: apps/wolfsshd/test/test_configuration.c [hunk 5]
```diff
@@ -153,12 +230,20 @@ int main(int argc, char** argv)
     (void)argc;
     (void)argv;
 
+    CleanupWildcardTest();
+    ret = SetupWildcardTest();
+    if (ret != 0) {
+        return 1;
```
Comment (haydenroche5) [inline apps/wolfsshd/test/test_configuration.c:236]: Prefer to return `WS_FATAL_ERROR` or another negative value for consistency.

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

File: apps/wolfsshd/configuration.c [hunk 6]
```diff
@@ -474,85 +475,133 @@ static int HandleInclude(WOLFSSHD_CONFIG *conf, const char *value)
...
+                else {
+                    WMEMCPY(path, ".", 1);
+                    path[1] = '\0';
+                    prefix = value;
+                    prefixLen = (int)(ptr - value);
+                }
+            }
+        }
+
+        if (!WOPENDIR(NULL, conf->heap, &d, path)) {
```
Comment (haydenroche5) [inline apps/wolfsshd/configuration.c:526]: Needs `if (ret == WS_SUCCESS)`.

File: apps/wolfsshd/configuration.c [hunk 7]
```diff
@@ -474,85 +475,133 @@ static int HandleInclude(WOLFSSHD_CONFIG *conf, const char *value)
...
+            WCLOSEDIR(&d);
         }
         else {
             /* Bad directory */
-            WFREE(filepath, NULL, 0);
-            return WS_BAD_ARGUMENT;
+            ret = WS_INVALID_PATH_E;
         }
-        WFREE(filepath, NULL, 0);
+        WFREE(path, conf->heap, DYNTYPE_PATH);
```
Comment (haydenroche5) [inline apps/wolfsshd/configuration.c:603]: Do  we need to check that path and filepath aren't NULL before doing these two `WFREE`s?

**Review by haydenroche5 — CHANGES_REQUESTED** [review]

**Review by haydenroche5 — APPROVED** [review]

**Review by haydenroche5 — APPROVED** [review]

**Pull request merged**